src/decoder/lattice-faster-decoder-combine-bucketqueue.h

```diff
@@ -0,0 +1,642 @@
+      links(links), next(next), in_queue(false) { }
+};
+
+template <typename Fst>
+struct BackpointerToken {
+  using ForwardLinkT = ForwardLink<BackpointerToken>;
+  using Token = BackpointerToken;
+  using StateId = typename Fst::Arc::StateId;
+
+  // BackpointerToken is like Token but also
+  // Standard token type for LatticeFasterDecoder.  Each active HCLG
+  // (decoding-graph) state on each frame has one token.
+
+  // tot_cost is the total (LM + acoustic) cost from the beginning of the
+  // utterance up to this point.  (but see cost_offset_, which is subtracted
+  // to keep it in a good numerical range).
+  BaseFloat tot_cost;
+
+  // exta_cost is >= 0.  After calling PruneForwardLinks, this equals
+  // the minimum difference between the cost of the best path, and the cost of
+  // this is on, and the cost of the absolute best path, under the assumption
+  // that any of the currently active states at the decoding front may
+  // eventually succeed (e.g. if you were to take the currently active states
+  // one by one and compute this difference, and then take the minimum).
+  BaseFloat extra_cost;
+
+  // Record the state id of the token
+  StateId state_id;
+
+  // 'links' is the head of singly-linked list of ForwardLinks, which is what we
+  // use for lattice generation.
+  ForwardLinkT *links;
+
+  //'next' is the next in the singly-linked list of tokens for this frame.
+  BackpointerToken *next;
+
+  // Best preceding BackpointerToken (could be a on this frame, connected to
+  // this via an epsilon transition, or on a previous frame).  This is only
+  // required for an efficient GetBestPath function in
+  // LatticeFasterOnlineDecoderTpl; it plays no part in the lattice generation
+  // (the "links" list is what stores the forward links, for that).
+  Token *backpointer;
+
+  // identitfy the token is in current queue or not to prevent duplication in
+  // function ProcessOneFrame().
+  bool in_queue;
+
+  inline void SetBackpointer (Token *backpointer) {
+    this->backpointer = backpointer;
+  }
+
+  inline BackpointerToken(BaseFloat tot_cost, BaseFloat extra_cost,
+                          StateId state_id, ForwardLinkT *links,
+                          Token *next, Token *backpointer):
+      tot_cost(tot_cost), extra_cost(extra_cost), state_id(state_id), 
+      links(links), next(next), backpointer(backpointer),
+      in_queue(false) { }
+};
+
+}  // namespace decoder
+
+
+template<typename Token>
+class BucketQueue {
+ public:
+  /** Constructor. 'cost_scale' is a scale that we multiply the token costs by
+   *  before intergerizing; a larger value means more buckets.
+   *  'best_cost_estimate' is an estimate of the best (lowest) cost  that
+   *  we are likely to encounter (e.g. the best cost that we have seen so far).
+   *  It is used to initialize 'bucket_storage_begin_'.
+   */
+  BucketQueue(BaseFloat best_cost_estimate, BaseFloat cost_scale = 1.0);
```

Comment:
Dan, how are you thinking about remove "best_cost_estimate" from the constructor of BucketQueue?